### ISSUE TITLE
64bit arguments passing with 68k .library ABI

### ISSUE THREAD

**Issue by chainq** [jens-maus/amissl#46]
We're trying to provide some support for AmiSSL in Free Pascal Compiler. This involves creating our own headers for AmiSSL, to talk to the ABI directly, based on the C includes. However, when it comes to the Motorola 68k/AmigaOS3 version, it looks like some of the C inline macros only specify a single register (say, `d0`) to pass `uint64_t` or `int64_t` parameters to the function. One simple example is `OPENSSL_init_ssl()`, where the flags parameter is `uint64_t`, yet if I check the inline macros, I only see `d0` mentioned. This obviously cannot work properly, however, in the `amissl_lib.sfd`, clearly `(d0-d1)` as register pair is specified to be used, but I can't see any trace of this in the inline macros.

So, what's the deal here? The `.sfd` file is ignored, the inline macros are wrong, did we find an issue, or it works regardless through some C compiler magic I we are not aware of? Thanks.

**Comment by Futaura** [comment]
@chainq Compiler magic :-)

Basically, for GCC/OS3, it is implicit that the register specified is the first of the pair if the type is 64-bit. I had to double-check this by disassembling as I couldn't remember, but I can see that GCC is correctly passing the high bits in d0 and the low bits in d1, for `OPENSSL_init_ssl()`. If you look at the `LP2()` macro itself, it puts the `uinit64_t` macro parameter into a variable of that type in register `d0` (and `d1` implicitly).  I don't think GCC has a way to explicitly specify `d0-d1`, as you would do in VBCC for example.

**Comment by chainq** [comment]
@Futaura thanks for the answers. Things are more clear now I think. We're going to add support to specify register pairs for 64bit values in Amiga-style `.library` calls in Free Pascal, so the AmiSSL headers can be completed.

Actually, meanwhile I learned, that one needs this for some of the double precision math libraries as well. We didn't support this, because Free Pascal provides its own soft-float facilities instead for FPU-less systems, so supporting the old and fairly limited math libs wasn't a priority. So yeah, that could then work too. Closing the ticket, thanks.
